File: 3200-3299/3286. Find a Safe Walk Through a Grid/Solution.cpp
// Approach-1 (0-1 BFS using Deque)
// T.C : O(m × n)
// S.C : O(m × n)

class Solution
{
public:
  vector<vector<int>> directions{
      {0, 1},
      {0, -1},
      {1, 0},
      {-1, 0}};

  bool findSafeWalk(vector<vector<int>> &grid, int health)
  {

    int rows = grid.size();

    int cols = grid[0].size();

    vector<vector<int>> cost(
        rows,
        vector<int>(cols, INT_MAX));

    deque<vector<int>> dequeQueue;

    int startCost = grid[0][0];

    cost[0][0] = startCost;

    dequeQueue.push_front(
        {0, 0, startCost});

    while (!dequeQueue.empty())
    {

      auto current = dequeQueue.front();

      dequeQueue.pop_front();

      int row = current[0];

      int col = current[1];

      int healthUsed = current[2];

      if (healthUsed > cost[row][col])
      {
        continue;
      }

      for (auto &direction : directions)
      {

        int nextRow = row + direction[0];

        int nextCol = col + direction[1];

        if (
            nextRow >= 0 &&
            nextRow < rows &&
            nextCol >= 0 &&
            nextCol < cols)
        {

          int cellCost = grid[nextRow][nextCol];

          int newCost = healthUsed + cellCost;

          if (newCost < cost[nextRow][nextCol])
          {

            cost[nextRow][nextCol] = newCost;

            if (cellCost == 0)
            {

              dequeQueue.push_front(
                  {nextRow, nextCol, newCost});
            }
            else
            {

              dequeQueue.push_back(
                  {nextRow, nextCol, newCost});
            }
          }
        }
      }
    }

    return cost[rows - 1][cols - 1] < health;
  }
};

/*
//Approach-2 (Dijkstra using Priority Queue)
//T.C : O(m × n log(m × n))
//S.C : O(m × n)

class Solution {
public:

    vector<vector<int>> directions{
        {0,1},
        {0,-1},
        {1,0},
        {-1,0}
    };

    bool findSafeWalk(vector<vector<int>>& grid, int health) {

        int rows = grid.size();

        int cols = grid[0].size();

        vector<vector<int>> cost(
            rows,
            vector<int>(cols, INT_MAX)
        );

        priority_queue<
            vector<int>,
            vector<vector<int>>,
            greater<vector<int>>
        > priorityQueue;

        int startCost = grid[0][0];

        cost[0][0] = startCost;

        priorityQueue.push(
            {startCost, 0, 0}
        );

        while (!priorityQueue.empty()) {

            auto current = priorityQueue.top();

            priorityQueue.pop();

            int healthUsed = current[0];

            int row = current[1];

            int col = current[2];

            if (healthUsed > cost[row][col]) {
                continue;
            }

            if (
                row == rows - 1 &&
                col == cols - 1
            ) {
                return healthUsed < health;
            }

            for (auto &direction : directions) {

                int nextRow = row + direction[0];

                int nextCol = col + direction[1];

                if (
                    nextRow >= 0 &&
                    nextRow < rows &&
                    nextCol >= 0 &&
                    nextCol < cols
                ) {

                    int newCost =
                        healthUsed +
                        grid[nextRow][nextCol];

                    if (
                        newCost <
                        cost[nextRow][nextCol]
                    ) {

                        cost[nextRow][nextCol] =
                            newCost;

                        priorityQueue.push(
                            {
                                newCost,
                                nextRow,
                                nextCol
                            }
                        );
                    }
                }
            }
        }

        return false;
    }
};

File: 3200-3299/3286. Find a Safe Walk Through a Grid/Solution.java
import java.util.*;

class Solution {

  int[][] dirs = { { 0, 1 }, { 0, -1 }, { 1, 0 }, { -1, 0 } };

  public boolean findSafeWalk(List<List<Integer>> grid, int health) {

    int m = grid.size();
    int n = grid.get(0).size();

    int[][] cost = new int[m][n];

    for (int[] row : cost) {
      Arrays.fill(row, Integer.MAX_VALUE);
    }

    Deque<int[]> deque = new ArrayDeque<>();

    int startCost = grid.get(0).get(0);

    cost[0][0] = startCost;

    deque.offerFirst(new int[] { 0, 0, startCost });

    while (!deque.isEmpty()) {

      int[] current = deque.pollFirst();

      int row = current[0];
      int col = current[1];
      int healthUsed = current[2];

      if (healthUsed > cost[row][col]) {
        continue;
      }

      for (int[] dir : dirs) {

        int nextRow = row + dir[0];
        int nextCol = col + dir[1];

        if (nextRow >= 0 && nextRow < m &&
            nextCol >= 0 && nextCol < n) {

          int cellCost = grid.get(nextRow).get(nextCol);

          int newCost = healthUsed + cellCost;

          if (newCost < cost[nextRow][nextCol]) {

            cost[nextRow][nextCol] = newCost;

            if (cellCost == 0) {
              deque.offerFirst(new int[] { nextRow, nextCol, newCost });
            } else {
              deque.offerLast(new int[] { nextRow, nextCol, newCost });
            }
          }
        }
      }
    }

    return cost[m - 1][n - 1] < health;
  }
}

/*
 * Approach 2 — Dijkstra (Priority Queue)
 * 
 * Time Complexity: O(m × n log(m × n))
 * 
 * Space Complexity: O(m × n)
 * 
 * import java.util.*;
 * 
 * class Solution {
 * 
 * int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};
 * 
 * public boolean findSafeWalk(List<List<Integer>> grid, int health) {
 * 
 * int m = grid.size();
 * int n = grid.get(0).size();
 * 
 * int[][] cost = new int[m][n];
 * 
 * for (int[] row : cost) {
 * Arrays.fill(row, Integer.MAX_VALUE);
 * }
 * 
 * PriorityQueue<int[]> pq =
 * new PriorityQueue<>((a,b) -> a[2] - b[2]);
 * 
 * int startCost = grid.get(0).get(0);
 * 
 * cost[0][0] = startCost;
 * 
 * pq.offer(new int[]{0,0,startCost});
 * 
 * while (!pq.isEmpty()) {
 * 
 * int[] current = pq.poll();
 * 
 * int row = current[0];
 * int col = current[1];
 * int healthUsed = current[2];
 * 
 * if (healthUsed > cost[row][col]) {
 * continue;
 * }
 * 
 * if (row == m - 1 && col == n - 1) {
 * return healthUsed < health;
 * }
 * 
 * for (int[] dir : dirs) {
 * 
 * int nextRow = row + dir[0];
 * int nextCol = col + dir[1];
 * 
 * if (nextRow >= 0 && nextRow < m &&
 * nextCol >= 0 && nextCol < n) {
 * 
 * int newCost =
 * healthUsed + grid.get(nextRow).get(nextCol);
 * 
 * if (newCost < cost[nextRow][nextCol]) {
 * 
 * cost[nextRow][nextCol] = newCost;
 * 
 * pq.offer(new int[]{
 * nextRow,
 * nextCol,
 * newCost
 * });
 * }
 * }
 * }
 * }
 * 
 * return false;
 * }
 * }
 * 
 */

/*
 * BFS using Deque (Optimal) — O(m × n)
 * Dijkstra using Priority Queue — O(m × n log(m × n))
 */